Filter outlier builds from the flake-report-creator jenkins report

dhiller opened the pull request:
According to the [three sigma rule](https://en.wikipedia.org/wiki/68%E2%80%9395%E2%80%9399.7_rule) we look at all builds that are deviating three times from the standard deviation of failures over all builds as outliers, and thus filter them from the report.

Besides that we create a `rate-build` command that can create a rating over all builds of a job over a certain period of time.

screenshot of the popup:
![image](https://user-images.githubusercontent.com/809335/184889704-b2b25b4f-c117-4285-a91b-9afa59c9d41e.png)

**Note: in the latest version I have truncated the mean and standard deviation values to two digits**

dhiller commented:
/retest

dhiller commented:
/test pull-project-infra-test-robots

dhiller commented:
/test pull-project-infra-test-robots

dhiller commented:
/uncc @davidvossel @tiraboschi @aglitke 
/cc @brianmcarey @xpivarc

fossedihelm commented:
/cc

fossedihelm reviewed (commented):
Thanks @dhiller! Great improvement e nice work!
Apart from some nits, which you are free to ignore, it all seems fair to me.
I found some difficulties in reviewing commit by commit as, sometimes, later commits were overwriting earlier commits. In this direction, probably the first commit could be totally included in the second. Anyway, great work.
/lgtm

dhiller commented on robots/cmd/badbuilds/main.go:
```diff
@@ -0,0 +1,105 @@
+		jLog.Fatalf("failed to get job %s: %v", opts.jobName, err)
+	}
+
+	startOfReport := time.Now().Add(-1 * opts.startFrom)
```
Sub can only use Duration: https://pkg.go.dev/time#Time.Sub

fossedihelm replied:
Oh my bad! Thanks. It is ok :+1:

dhiller commented on robots/cmd/flake-report-creator/cmd/jenkins.go:
```diff
@@ -347,16 +387,20 @@ func runJenkinsReport(cmd *cobra.Command, args []string) error {
 
 	startOfReport := time.Now().Add(-1 * opts.startFrom)
-	endOfReport := time.Now()
+	startOfReportForRatings := time.Now().Add(-1 * opts.startFromForRatings)
```
Sub requires Duration: https://pkg.go.dev/time#Time.Sub

dhiller commented on robots/pkg/flakefinder/build/rating.go:
```diff
@@ -0,0 +1,127 @@
+	mean := float64(totalFailures) / float64(totalCompletedBuilds)
+
+	sumOfSquareDeviations := calculateSumOfSquareDeviations(buildNumbersToFailures, mean)
+	variance := float64(sumOfSquareDeviations) / float64(totalCompletedBuilds)
```
Thanks, good catch!

dhiller commented on robots/cmd/badbuilds/README.md:
```diff
@@ -0,0 +1,24 @@
+badbuilds
+=========
+
+Tool to fetch three sigma ratings over a set of builds for a jenkins job.
```
Thanks, good idea!

dhiller commented:
> Thanks @dhiller! Great improvement e nice work! Apart from some nits, which you are free to ignore, it all seems fair to me. I found some difficulties in reviewing commit by commit as, sometimes, later commits were overwriting earlier commits. In this direction, probably the first commit could be totally included in the second. Anyway, great work. /lgtm

Thank you for the review @fossedihelm! I've updated the PR, I've merged the first two commits as requested, added some info in the README, removed the redundant cast, and fixed a bazel target name. All changes in the last commit. PTAL, thank you!

fossedihelm commented:
Thanks @dhiller! Looks great!
Just for consistency, can you adjust the PR description since the command was renamed from `badbuild` to `rate-build`?
Thanks
/lgtm

dhiller commented:
> Thanks @dhiller! Looks great! Just for consistency, can you adjust the PR description since the command was renamed from `badbuild` to `rate-build`? Thanks /lgtm

Great catch again, thanks!

dhiller reviewed (commented):
/approve
